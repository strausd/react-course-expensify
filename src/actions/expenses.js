import uuid from 'uuid';

import database from '../firebase/firebase';

export const addExpense = (expense) => {
    return {
        type: 'ADD_EXPENSE',
        expense
    };
};

export const startAddExpense = (expenseData = {}) => {
    return (dispatch) => {
        const {
            description = '',
            note = '',
            amount = 0,
            createdAt = 0
        } = expenseData;

        const expense = { description, note, amount, createdAt };

        return database.ref('expenses').push(expense).then((ref) => {
            dispatch(addExpense({
                id: ref.key,
                ...expense
            }));
        }).catch(() => {});
    }
};

export const removeExpense = ({ id }) => {
    return {
        type: 'REMOVE_EXPENSE',
        id
    };
};

export const editExpense = (id, updates) => {
    return {
        type: 'EDIT_EXPENSE',
        id,
        updates
    };
};

export const setExpenses = (expenses) => {
    return {
        type: 'SET_EXPENSES',
        expenses
    }
};

export const startSetExpenses = () => {
    return (dispatch) => {
        return database.ref('expenses').once('value').then((dataSnapshot) => {
            const expenses = [];
            dataSnapshot.forEach((childSnapshot) => {
                const expenseData = childSnapshot.val();
                expenses.push({
                    id: childSnapshot.key,
                    ...expenseData
                });
            });
            dispatch(setExpenses(expenses));
        }).catch(() => {});
    };
};