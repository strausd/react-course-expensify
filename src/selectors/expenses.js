import moment from 'moment';

// Get visible expenses
export const getVisibleExpenses = (expenses, { text, sortBy, startDate, endDate }) => {
    return expenses.filter((expense) => {
        const createdAtMoment = moment(expense.createdAt);
        const startDateMatch = startDate ? startDate.isSameOrBefore(createdAtMoment) : true;
        const endDateMatch = endDate ? endDate.isSameOrAfter(createdAtMoment): true;
        const textMatch = !text || expense.description.toLowerCase().includes(text.toLowerCase());

        return startDateMatch && endDateMatch && textMatch;
    }).sort((a, b) => {
        if (sortBy === 'date') {
            return a.createdAt < b.createdAt ? 1 : -1;
        } else if (sortBy === 'amount') {
            return a.amount < b.amount ? 1 : -1;
        }
    });
};

export const getExpensesTotal = (expenses = []) => {
    let total = 0;
    expenses.forEach(expense => total += expense.amount);
    return total;
};
