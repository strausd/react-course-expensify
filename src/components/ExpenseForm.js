import React from 'react';
import moment from 'moment';
import { connect } from 'react-redux';
import { SingleDatePicker } from 'react-dates';

import { removeExpense } from '../actions/expenses';


export class ExpenseForm extends React.Component {
    constructor(props) {
        super(props);
        this.state = {
            description: props.expense ? props.expense.description : '',
            note: props.expense ? props.expense.note : '',
            amount: props.expense ? (props.expense.amount/100).toString() : '',
            createdAt: props.expense ? moment(props.expense.createdAt) : moment(),
            calendarFocused: false,
            error: '',
        };
        this.onDescriptionChange = this.onDescriptionChange.bind(this);
        this.onNoteChange = this.onNoteChange.bind(this);
        this.onAmountChange = this.onAmountChange.bind(this);
        this.onDateChange = this.onDateChange.bind(this);
        this.onFocusChange = this.onFocusChange.bind(this);
        this.onSubmit = this.onSubmit.bind(this);
        this.displayErrorMessage = this.displayErrorMessage.bind(this);
    }

    onDescriptionChange(e) {
        const description = e.target.value;
        this.setState(() => ({ description }));
    }

    onNoteChange(e) {
        const note = e.target.value;
        this.setState(() => ({ note }));
    }

    onAmountChange(e) {
        const amount = e.target.value;
        if (!amount || amount.match(/^\d+(\.\d{0,2})?$/)) {
            this.setState(() => ({ amount }));
        }
    }

    onDateChange(createdAt) {
        this.setState(() => ({ createdAt }));
    }

    onFocusChange({ focused }) {
        this.setState(() => ({
            calendarFocused: focused
        }));
    }

    onSubmit(e) {
        e.preventDefault();

        if (this.state.description && this.state.amount ) {
            this.setState(() => ({error: ''}));
            this.props.onSubmit({
                description: this.state.description,
                amount: Math.round(parseFloat(this.state.amount) * 100),
                createdAt: this.state.createdAt.valueOf(),
                note: this.state.note
            });
        } else {
            this.setState(() => ({error: 'Please provide description and amount.'}));
        }
    }

    displayErrorMessage() {
        if (this.state.error) {
            return (
                <p>{this.state.error}</p>
            );
        } else {
            return undefined;
        }
    }

    render() {
        return (
            <div>
                {this.displayErrorMessage()}
                <form
                    onSubmit={this.onSubmit}
                >
                    <input
                        type="text"
                        placeholder="Description"
                        autoFocus
                        value={this.state.description}
                        onChange={this.onDescriptionChange}
                    />
                    <input
                        type="text"
                        placeholder="Amount"
                        value={this.state.amount}
                        onChange={this.onAmountChange}
                    />
                    <SingleDatePicker
                        date={this.state.createdAt}
                        onDateChange={this.onDateChange}
                        focused={this.state.calendarFocused}
                        onFocusChange={this.onFocusChange}
                        numberOfMonths={1}
                        isOutsideRange={() => false}
                        showClearDate={true}
                    />
                    <textarea
                        placeholder="Add a note for your expense (optional)"
                        value={this.state.note}
                        onChange={this.onNoteChange}
                    >
                    </textarea>
                    <button>Save Expense</button>
                </form>
            </div>
        );
    }
}

export default connect()(ExpenseForm);
