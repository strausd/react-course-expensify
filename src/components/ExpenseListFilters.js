import React from 'react';
import { connect } from 'react-redux';
import { DateRangePicker } from 'react-dates';

import { setTextFilter, sortByAmount, sortByDate, setStartDate, setEndDate } from '../actions/filters';


export class ExpenseListFilters extends React.Component {
    constructor(props) {
        super(props);
        this.state = {
            calendarFocused: null
        };
        this.onDatesChange = this.onDatesChange.bind(this);
        this.onFocusChange = this.onFocusChange.bind(this);
        this.onTextChange = this.onTextChange.bind(this);
        this.onSortByChange = this.onSortByChange.bind(this);
    }

    onDatesChange({ startDate, endDate }) {
        this.props.setStartDate(startDate);
        this.props.setEndDate(endDate);
    }

    onFocusChange(calendarFocused) {
        this.setState(() => ({ calendarFocused }));
    }

    onTextChange(e) {
        this.props.setTextFilter(e.target.value);
    }

    onSortByChange(e) {
        if (e.target.value === 'amount') {
            this.props.sortByAmount();
        } else if (e.target.value === 'date') {
            this.props.sortByDate();
        }
    }

    render() {
        return (
            <div>
                <input
                    type="text"
                    value={this.props.filters.text}
                    onChange={this.onTextChange}
                />
                <select
                    name="sortBy"
                    value={this.props.filters.sortBy}
                    onChange={this.onSortByChange}
                >
                    <option value="date">Date</option>
                    <option value="amount">Amount</option>
                </select>
                <DateRangePicker
                    startDate={this.props.filters.startDate}
                    endDate={this.props.filters.endDate}
                    onDatesChange={this.onDatesChange}
                    focusedInput={this.state.calendarFocused}
                    onFocusChange={this.onFocusChange}
                    numberOfMonths={1}
                    isOutsideRange={() => false}
                    showClearDates={true}
                />
            </div>
        );
    }
}

const mapStateToProps = (state) => {
    return {
        filters: state.filters
    };
};

const mapDispatchToProps = (dispatch, props) => {
    return {
        setTextFilter: (text) => dispatch(setTextFilter(text)),
        setStartDate: (startDate) => dispatch(setStartDate(startDate)),
        setEndDate: (endDate) => dispatch(setEndDate(endDate)),
        sortByAmount: () => dispatch(sortByAmount()),
        sortByDate: () => dispatch(sortByDate())
    };
};

export default connect(mapStateToProps, mapDispatchToProps)(ExpenseListFilters);
