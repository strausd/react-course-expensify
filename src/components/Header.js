import React from 'react';
import { NavLink, withRouter } from 'react-router-dom';
import { connect } from 'react-redux';

import { startLogout } from "../actions/auth";

export const Header = ({ startLogout }) => (
    <header>
        <h1>Expensify</h1>
        <NavLink activeClassName="is-active" to="/dashboard">Dashboard</NavLink>
        <NavLink activeClassName="is-active" to="/create">Create Expense</NavLink>
        <button onClick={ startLogout }>Logout</button>
    </header>
);

const mapDispatchToProps = (dispatch) => {
    return {
        startLogout: () => dispatch(startLogout())
    };
}

export default withRouter(connect(undefined, mapDispatchToProps)(Header));
