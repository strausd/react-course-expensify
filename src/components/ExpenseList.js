import React from 'react';
import { connect } from 'react-redux';

import ExpenseListItem from './ExpenseListItem';
import { getVisibleExpenses } from '../selectors/expenses';
import { startSetExpenses } from '../actions/expenses';

export class ExpenseList extends React.Component {
    constructor(props) {
        super(props);
        props.startSetExpenses();
    }

    render() {
        return (
            <div>
                {
                    this.props.expenses.length === 0 ? (
                        <p>No expenses</p>
                    ) : (
                            this.props.expenses.map(expense => <ExpenseListItem key={expense.id} {...expense} />)
                        )
                }
            </div>
        );
    }
}

const mapStateToProps = (state) => {
    return {
        expenses: getVisibleExpenses(state.expenses, state.filters)
    };
};

const mapDispatchToProps = (dispatch) => {
    return {
        startSetExpenses: () => dispatch(startSetExpenses())
    };
};

export default connect(mapStateToProps, mapDispatchToProps)(ExpenseList);
