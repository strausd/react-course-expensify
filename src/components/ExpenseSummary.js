import React from 'react';
import { connect } from 'react-redux';
import numeral from 'numeral';

import { getExpensesTotal, getVisibleExpenses } from '../selectors/expenses';


export class ExpenseSummary extends React.Component {
    constructor(props) {
        super(props);
        this.getFormattedExpenseTotal = this.getFormattedExpenseTotal.bind(this);
        this.getFinalString = this.getFinalString.bind(this);
    }

    getFormattedExpenseTotal() {
        const amount = this.props.getExpensesTotal(this.props.getVisibleExpenses());
        return numeral(amount / 100).format('$0,0.00');
    }

    getFinalString() {
        const num_expenses = this.props.getVisibleExpenses().length;
        const amount = this.getFormattedExpenseTotal();
        const expense_string = num_expenses === 1 ? 'expense' : 'expenses';
        return (
            <p>Viewing {num_expenses} {expense_string} totalling {amount}</p>
        );
    }

    render() {
        return (
            <div>
                {this.getFinalString()}
            </div>
        );
    }
}

const mapStateToProps = (state, props) => {
    return {
        getVisibleExpenses: () => getVisibleExpenses(state.expenses, state.filters),
        getExpensesTotal: (expenses) => getExpensesTotal(expenses)
    };
};

export default connect(mapStateToProps)(ExpenseSummary);