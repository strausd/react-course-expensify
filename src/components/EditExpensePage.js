import React from 'react';
import { connect } from 'react-redux';

import ExpenseForm from './ExpenseForm';
import { editExpense, removeExpense } from '../actions/expenses';

export class EditExpensePage extends React.Component {
    constructor(props) {
        super(props);
        this.onSubmit = this.onSubmit.bind(this);
        this.onDeleteExpense = this.onDeleteExpense.bind(this);
    }

    onSubmit(expense) {
        this.props.editExpense(this.props.expense.id, expense);
        this.props.history.push('/');
    }

    onDeleteExpense(e) {
        this.props.removeExpense({ id: this.props.expense.id });
        this.props.history.push('/');
    }

    render() {
        return (
            <div>
                <ExpenseForm
                    expense={this.props.expense}
                    onSubmit={this.onSubmit}
                />
                <button onClick={this.onDeleteExpense}>
                    Delete Expense
                </button>
            </div>
        );
    }
};

const mapStateToProps = (state, props) => {
    return {
        expense: state.expenses.find((expense) => {
            return expense.id === props.match.params.id;
        })
    }
};

const mapDispatchToProps = (dispatch) => {
    return {
        editExpense: (id, expense) => dispatch(editExpense(id, expense)),
        removeExpense: ({ id }) => dispatch(removeExpense({ id }))
    };
};

export default connect(mapStateToProps, mapDispatchToProps)(EditExpensePage);
