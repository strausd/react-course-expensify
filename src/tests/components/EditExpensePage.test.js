import React from 'react';
import { shallow } from 'enzyme';

import expenses from '../fixtures/expenses';
import { EditExpensePage } from '../../components/EditExpensePage';


let wrapper, editExpense, removeExpense, expense, history;

beforeAll(() => {
    editExpense = jest.fn();
    removeExpense = jest.fn();
    history = { push: jest.fn() }
    expense = expenses[0];
    wrapper = shallow(<EditExpensePage editExpense={editExpense} removeExpense={removeExpense} history={history} expense={expense} />);
});

test('should render EditExpensePage correctly', () => {
    expect(wrapper).toMatchSnapshot();
});

test('should verify onSubmit calls functions with proper data', () => {
    const component_expense = wrapper.find('Connect(ExpenseForm)').prop('expense');
    expect(component_expense).toEqual(expense);
    wrapper.find('Connect(ExpenseForm)').prop('onSubmit')(expense);
    expect(history.push).toHaveBeenLastCalledWith('/');
    expect(editExpense).toHaveBeenLastCalledWith(expense.id, expense);
});

test('should verify onDeleteExpense calls function with proper data', () => {
    wrapper.find('button').simulate('click');
    expect(history.push).toHaveBeenLastCalledWith('/');
    expect(removeExpense).toHaveBeenLastCalledWith({ id: expense.id });
});