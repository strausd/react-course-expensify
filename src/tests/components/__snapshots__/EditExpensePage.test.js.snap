// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`should render EditExpensePage correctly 1`] = `
<div>
  <Connect(ExpenseForm)
    expense={
      Object {
        "amount": 195,
        "createdAt": 1514181600000,
        "description": "Gum",
        "id": "1",
        "note": "",
      }
    }
    onSubmit={[Function]}
  />
  <button
    onClick={[Function]}
  >
    Delete Expense
  </button>
</div>
`;
